### DISCUSSION TITLE
A new method for the drawing API: curve offsets

### DISCUSSION THREAD

**Discussion by sgenoud** [sgenoud/replicad#65] [Announcements]
This one was more difficult than I thought. I have added a `offset` method on drawings.

What does that mean? You can now take a 2d drawing a create an offset (i.e. contour). You can even make many of them:

<img width="561" alt="Capture d’écran 2022-12-29 à 11 54 49" src="https://user-images.githubusercontent.com/263325/209943321-47ef962c-2540-4d76-8da3-ed6a930f9321.png">

It works as you would expect:

```js
function main() {
  const shape = draw([10, 15])
    .hLine(-5)
    .vLine(5)
    .hLine(20)
    .vLine(-30)
    .hSagittaArc(-15, 3)
    .vLine(5)
    .hSagittaArc(5, -2)
    .close();

  return [
    { shape, color: "black", name: "Base Shape" },
    { shape: shape.offset(-2), color: "blue", name: "Inner offset" },
    { shape: shape.offset(2), color: "red", name: "Outer offset" },
  ];
}
```
Which draws

<img width="470" alt="Capture d’écran 2022-12-29 à 12 15 37" src="https://user-images.githubusercontent.com/263325/209943945-355b37f2-ef79-481f-8018-4d36e3bac91e.png">

Note that there are some known bugs around bezier curves that I need to fix (and I am curious to see how you break it as well!

**Comment by sgenoud** [comment]
Some more! The logic for offsetting lines was laying the ground work for: fillet and chamfer in 2D. And I went all the way to expose it with two different API.

**While drawing**

```js
const shape = draw().hLine(10).customCorner(2, 'fillet').vLine(5).close()
```

**on a closed drawing**

```js
const shape = draw()
        .hLine(10)
        .vLine(5)
        .close()
        .fillet(2, (e) => e.ofAngle(90))
```

The first example is nicer when you are drawing a full shape, and the second can be used after some boolean operations have been done, for instance:

```js
  const shape = drawRoundedRectangle(10, 20)
    .fuse(drawCircle(5).translate(5, 10))
    .fuse(drawCircle(3).translate(-3.5, -11));

  const customWithFinder = shape.fillet(2, (e) => e.ofAngle(90));

  return [
    {
      shape: shape,
      name: "Base Shape",
    },
    {
      shape: customWithFinder.translate(20, 0),
      name: "Modified with selectors",
      color: "blue",
    },
  ];
```

will show something like this

<img width="854" alt="Capture d’écran 2023-01-05 à 11 51 39" src="https://user-images.githubusercontent.com/263325/210763287-7c918caa-ff77-40af-bd09-be29b6ae6c43.png">

**Reply by Papooch** [reply]
Shouldn't there be fillets around the smaller circle as well?

**Reply by sgenoud** [reply]
they are not 90 degrees!
